### PR TITLE
Implement MovementSounds.

### PR DESCRIPTION
**Pull request by phrohdoh**
Works for `Mobile` actors.
Aircraft are much more complicated and will need quite a bit of work.

Depends on https://github.com/OpenRA/OpenRA/pull/7871.

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/Mobile.cs [hunk 1]
```diff
@@ -66,6 +66,15 @@ public class MobileInfo : IMobileInfo, IOccupySpaceInfo, IFacingInfo, UsesInit<F
 		[Desc("Can the actor be ordered to move in to shroud?")]
 		public readonly bool MoveIntoShroud = true;
 
+		[Desc("Randomly select a sound from this list to play when this actor beings moving.")]
+		public readonly string[] Sounds = { };
+
+		[Desc("Should allied players hear movement sounds?")]
+		public readonly bool AlliesHearSounds = true;
+
+		[Desc("Should enemy players hear movement sounds?")]
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Mobile.cs]: Actually it seems like setting this to true will also make the sound hearable for neutrals.

Comment (penev92) [inline OpenRA.Mods.Common/Traits/Mobile.cs]: If it is engine sounds it doesn't even make sense to restrict it at all.

Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Mobile.cs]: It does for configurability.

**Comment by Phrohdoh** [comment]
Maybe this should be moved into its own trait, actually.
So you can have different sounds depending on player and other conditions.

File: OpenRA.Mods.Common/Traits/Mobile.cs [hunk 2]
```diff
@@ -66,6 +66,15 @@ public class MobileInfo : IMobileInfo, IOccupySpaceInfo, IFacingInfo, UsesInit<F
 		[Desc("Can the actor be ordered to move in to shroud?")]
 		public readonly bool MoveIntoShroud = true;
 
+		[Desc("Randomly select a sound from this list to play when this actor beings moving.")]
```
Comment (WolfGaming) [inline OpenRA.Mods.Common/Traits/Mobile.cs]: just as a question but how does this sentence make sense?
"when this actor beings moving."

Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Mobile.cs]: I'm not sure what you're really asking here.

<action> done at <time>, or, <action> done under <condition>.

File: OpenRA.Mods.Common/Traits/MovementSounds.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
+﻿// #region Copyright & License Information
+// /*
+//  * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+//  * This file is part of OpenRA, which is free software. It is made
+//  * available to you under the terms of the GNU General Public License
+//  * as published by the Free Software Foundation. For more information,
+//  * see COPYING.
+//  */
+// #endregion
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/MovementSounds.cs]: We should probably file a patch towards @MonoDevelop to fix this annoyance.

Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/MovementSounds.cs]: Good idea.

File: OpenRA.Mods.Common/Traits/MovementSounds.cs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+using System;
+using OpenRA.Traits;
+using OpenRA.Support;
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/MovementSounds.cs]: Don't forget to fix `[SA1210] Using directives must be sorted alphabetically by the namespaces.` when you do the rebase.

**Comment by Mailaender** [comment]
https://github.com/OpenRA/OpenRA/pull/7871 has been merged.

**Comment by Phrohdoh** [comment]
Rebased.

**Comment by Mailaender** [comment]
It doesn't come with a use-case. Is this needed for RA2? If so, could you link a branch it can be tested against?

**Comment by penev92** [comment]
Yeah, RA2 units have such sounds.

**Comment by Phrohdoh** [comment]
@bvandenbogaard Which units need this?

**Comment by GraionDilach** [comment]
If this is about MoveStart sounds then all vehicles excluding the Chrono Miner had such in RA2.

File: OpenRA.Mods.Common/Traits/MovementSounds.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+#endregion
+
+using OpenRA.Support;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public class MovementSoundsInfo : ITraitInfo
+	{
+		[Desc("Randomly select a sound from this list to play when this actor beings moving.")]
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/MovementSounds.cs]: "begins", or "starts"

File: OpenRA.Mods.Common/Traits/MovementSounds.cs [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+		public readonly string[] Sounds = { };
+
+		[Desc("Should the selected sound loop?")]
+		public readonly bool LoopSound = true;
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/MovementSounds.cs]: I believe the default that RA2 needs is `false` ?

Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/MovementSounds.cs]: You believe, I confirm. RA2 sound.ini had a property called loop for that which wasn't default at all.

File: OpenRA.Mods.Common/Traits/MovementSounds.cs (hunk 5 above)
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/MovementSounds.cs]: Oops.

**Comment by obrakmann** [comment]
Just as an idea: would tracking the state of `IMove::IsMoving` to start/stop the sound work for this as well? That would then work for planes, too, if that is needed.

**Comment by Phrohdoh** [comment]
I could do that but then that check would need to be ran each tick.

File: OpenRA.Mods.Common/Traits/Mobile.cs [hunk 7]
```diff
@@ -304,6 +304,12 @@ public bool CanEnterCell(World world, Actor self, CPos cell, out int movementCos
 		bool IOccupySpaceInfo.SharesCell { get { return SharesCell; } }
 	}
 
+	public interface INotifyMovement
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/Mobile.cs]: Please move this to `TraitsInterfaces.cs`, as this file is horrible as it is, with two monsterous classes, an enum and nested classes :X

**Comment by pchote** [comment]
Are you sure you can't use the existing `IMove.IsMoving` for this.  There are cases (e.g. Drag, the TD harvester docking, everything aircraft related) that will cause actor movement without triggering the interface.

**Comment by Phrohdoh** [comment]
I said above that I could do that, but then I'd have to check each tick for a change to call my interface methods.

**Comment by Unit158** [comment]
It is now july. We should close this if you are not going to work on it further.

**Pull request closed without merging**